bqpump-core/src/main/java/ai/aliz/bqpump/notification/PubsubUtil.java

```
@@ -37,16 +39,30 @@
     public final String ROLES_PUBSUB_PUBLISHER = "roles/pubsub.publisher";
     
     private final Integer ACK_DEADLINE_SECONDS = 10;
-
+    
     @SneakyThrows
     public void publishAsJson(ProjectTopicName topic, Object gsonableData) {
         log.info("Publishing message: {} to topic {}", gsonableData, topic);
         final String jsonData = GSON.toJson(gsonableData);
         final PubsubMessage message = PubsubMessage.newBuilder().setData(ByteString.copyFromUtf8(jsonData)).build();
         final Publisher publisher = Publisher.newBuilder(topic).build();
-        publisher.publish(message);
+        
+        ApiFutures.addCallback(publisher.publish(message), new ApiFutureCallback<String>() {
+            @Override
+            public void onFailure(Throwable t) {
+                log.error("Failed to send message to PubSub.", t);
```

why don't you call shutdown if it fails?

---

btw I'm not sure we should swallow the exception here at all.... IMO it should be thrown out so the appengine task queue can retry the task

---

I'm not sure if it's appengine retrying sending the message at all. I thought if PubSub couldn't send the message then it can be configured to retry AND it calls this callback.

---

"I'm not sure if it's appengine retrying sending the message at all. "
This method publishes a message to the pubsub (sends a message to a pubsub topic)... and then any further notification to any subscription is handled - and retried - there, not here... but code itself runs on GAE (```BigqueryQueryDeferredTask```) so this failure is all about failing to put the message into a topic, so if it fails appengine will retry it... if you swallow the exception, and it fails, the message will never arrive into the pubsub topic, any none of the subscriptions will get it...

"Anyway, anything logged here asynchronously is not logged anyway"
That most likely happens because it gets executed after the request has been finished, there is no logging limitation in threads regarding this on GAE.

"Anyway, I got the example from https://cloud.google.com/pubsub/docs/quickstart-client-libraries"
Check https://cloud.google.com/pubsub/docs/publisher then. Seems better in this case. Don't forget that the example you used is a simple app with a main method, so it runs until there is a non-daemon thread running... meanwhile at GAE if the request finishes, it kills of every "request thread".

Btw even that example contains a ```.shutdown()``` call for both cases ;)

---

Yes, I've added a try-finally block and shut down the publisher there.

I think I get what you're saying now, any failure when sending a message should be handled by the ``BigQueryDeferredTask``. In that case, I'll of course have to await the call to PubSub and throw an exception to cause the DeferredTask to be retried by App Engine. I just though calling ``Publisher#publish`` started some "internal" PubSub async task that gets executed by some PubSub-specific internal async executor which handles failures and calls the defined async methods. I couldn't find any docs on this.

---

I don't get it why do you think this. It executes it inside a thread - threadpool to be precise -:
https://github.com/GoogleCloudPlatform/google-cloud-java/blob/71637ce831ced5208af752d4e02bafeb4212e51c/google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java#L195-L272
... but that can't do appengine deferred task-level retry... it can do internal retries, and some exponential backoff, but that's all. It cannot trigger a complete task retry.

---

Thanks for the clarification, I've now implemented the synchronized version.